src/packet/packet_reader.rs

```diff
@@ -0,0 +1,127 @@
+use crate::net::constants::STANDARD_HEADER_SIZE;
+use crate::packet::header::{
+    AckedPacketHeader, ArrangingHeader, FragmentHeader, HeaderReader, StandardHeader,
+};
+use crate::{IntoBoxedSlice, Result};
+
+use std::io::Cursor;
+
+/// Could be used to read the packet contents of laminar.
+///
+/// # Remarks
+/// - `PacketReader` is using an underlying `Cursor` to manage the reading of the bytes.
+/// - `PacketReader` can interpret where some data is located in the buffer, that's why you don't have to worry about the position of the `Cursor`.
+pub struct PacketReader<'s> {
+    buffer: &'s [u8],
+    cursor: Cursor<&'s [u8]>,
+}
+
+impl<'s> PacketReader<'s> {
+    /// Construct a new instance of `PacketReader`, the given `buffer` will be used to read information from.
+    pub fn new(buffer: &'s [u8]) -> PacketReader<'s> {
+        PacketReader {
+            buffer,
+            cursor: Cursor::new(buffer),
+        }
+    }
+
+    /// Read the `StandardHeader` from the underlying buffer.
+    ///
+    /// # Remark
+    /// - Will change the position to the location of `StandardHeader`
+    pub fn read_standard_header(&mut self) -> Result<StandardHeader> {
+        self.cursor.set_position(0);
+        StandardHeader::read(&mut self.cursor)
+    }
+
+    /// Read the `StandardHeader` from the underlying buffer.
+    ///
+    /// # Remark
+    /// - Will change the position to the location of `StandardHeader`
+    pub fn read_arranging_header(&mut self, start_offset: u16) -> Result<ArrangingHeader> {
```

Comment:
I worry that these methods don't have a check to see if it's possible that this particular buffer should have an `ArrangingHeader` before return an `ArrangingHeader` with junk bytes it read off of the slice.

Reply:
Right, let me make some validation methods over there.

Reply:
Think of it this way, if I send you a bunch of garbage bytes over the socket, I would want the reader to error if the buffer doesn't represent a properly formatted laminar packet. I'm less worried about the case in which we run out of bytes to read. More that those bytes may not be what we expect and we should protect against that at the reader level.